fud2/fud-core/src/exec/plan/egg_planner.rs

```diff
@@ -0,0 +1,260 @@
+
+    // Collect states into a pattern.
+    let mut rhs: PatternAst<StateLanguage> = Default::default();
+    let state_ids = all_states
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            if op.output.contains(s) {
+                rhs.add(egg::ENodeOrVar::ENode(StateLanguage::C))
+            } else if op.input.contains(s) {
+                rhs.add(egg::ENodeOrVar::ENode(StateLanguage::X))
+            } else {
+                rhs.add(egg::ENodeOrVar::Var(
+                    format!("?s{}", i).parse().unwrap(),
+                ))
+            }
+        })
+        .collect::<Vec<_>>();
+    let state_pattern = rhs.add(egg::ENodeOrVar::ENode(StateLanguage::States(
+        state_ids.into_boxed_slice(),
+    )));
+
+    // Collect ops into a pattern.
+    let op_ids = all_ops
+        .iter()
+        .enumerate()
+        .map(|(i, &o)| {
+            if through.contains(&op_ref) && o == op_ref {
+                rhs.add(egg::ENodeOrVar::ENode(StateLanguage::C))
+            } else {
+                rhs.add(egg::ENodeOrVar::Var(
+                    format!("?o{}", i).parse().unwrap(),
+                ))
+            }
+        })
+        .collect::<Vec<_>>();
+    let op_pattern = rhs.add(egg::ENodeOrVar::ENode(StateLanguage::Ops(
+        op_ids.into_boxed_slice(),
+    )));
+
+    rhs.add(egg::ENodeOrVar::ENode(StateLanguage::Root([
+        state_pattern,
+        op_pattern,
+    ])));
+    let rhs = Pattern::new(rhs);
+
+    rewrite!(name; lhs => rhs)
+}
+
+impl FindPlan for EggPlanner {
+    fn find_plan(
+        &self,
+        start: &[StateRef],
+        end: &[StateRef],
+        through: &[OpRef],
+        ops: &PrimaryMap<OpRef, Operation>,
+        states: &PrimaryMap<StateRef, State>,
+    ) -> Option<Vec<Step>> {
+        // Collect all ops and states into sorted `Vec`s.
+        let all_states: Vec<_> = states.keys().collect();
+        let all_ops: Vec<_> = ops.keys().collect();
+
+        // Construct egg rewrites for each op.
+        let rules: Vec<Rewrite<StateLanguage, ()>> = ops
+            .iter()
+            .map(|(op_ref, op)| {
+                rewrite_from_op(op_ref, op, through, &all_states, &all_ops)
+            })
+            .collect();
+
+        // Construct initial expression.
+        let start_expr: RecExpr<StateLanguage> =
+            language_expr(start, &[], &all_states, &all_ops);
+
+        // Find a solution.
+        let mut runner = Runner::default()
+            .with_explanations_enabled()
+            .with_expr(&start_expr)
+            .run(&rules);
+
+        // Create solution expression. This assumes that the ops generate exactly the requested
+        // files with no extras.
+        let end_expr: RecExpr<StateLanguage> =
+            language_expr(end, through, &all_states, &all_ops);
+
+        // If the end expression exists, retrieve it using steps.
+        if runner.egraph.lookup_expr(&end_expr).is_some() {
```

Comment:
Here's a suggestion for how to rewrite this to be less nested. Up to you if you think it's cleaner.

```rust
        runner
            .egraph
            .lookup_expr(&end_expr)
            .map(|_| {
                runner
                    .explain_equivalence(&start_expr, &end_expr)
                    .make_flat_explanation()
                    .iter()
                    // only keep terms with forward rules
                    .filter_map(|t| t.forward_rule)
                    // re-parse the op's reference
                    .map(|r| {
                        OpRef::from_u32(r.to_string().parse::<u32>().unwrap())
                    })
                    .map(|op_ref| (op_ref, ops[op_ref].output.to_vec()))
                    .collect::<Vec<_>>()
            })
            .filter(|steps| !steps.is_empty())
```